### ISSUE TITLE
Reconsider `Entry` type, dlopen use

### ISSUE THREAD

**Issue by Ralith** [MaikKlein/ash#68]
The standard way to use Vulkan is to link libvulkan directly, and then use the directly-exported `vkGet{Instance|Device}ProcAddr` functions to load function pointers as needed. Using dlopen instead of linking complicates ash's public API by requiring the `Entry` type, and makes it more difficult to use ash-based programs in environments where dynamic libraries are not guaranteed to be in `LD_LIBRARY_PATH`. It's unclear what benefit is gained in exchange. If there is indeed a sufficiently large benefit to justify the increased friction, it should be documented to prevent this question from recurring.

**Comment by GabrielMajeri** [comment]
> The standard way to use Vulkan is to link libvulkan directly

I disagree. In that case, if the user does not have Vulkan installed, then the game will fail to launch. Especially on mobile devices, where driver updates are quite rare, many games prefer to try to dynamically load Vulkan, and fall back to OpenGL if that's not supported.

> makes it more difficult to use ash-based programs in environments where dynamic libraries are not guaranteed to be in LD_LIBRARY_PATH

Perhaps in those cases you could link the Vulkan loader into your program, or include it as a shared object with your app and link against it. In the end, loading the drivers on any platform will require dynamic linking.

I do agree that there should be an option to choose between static linking and dynamic linking with `Entry`.

**Comment by MaikKlein** [comment]
dlopen is pretty much the standard way of linking Vulkan and I don't see any negative effects. And yes I want to add a better way of choosing the library. A year ago it was just not possible to make it configurable on stable, but I think this has changed.

**Comment by GabrielMajeri** [comment]
@MaikKlein We could do something like this:

```rust
impl Entry {
  pub fn dynamic() -> Self {
    // use dlopen
  }

  pub fn static() -> Self {
    #[link(name = "vulkan")]
    extern "system" {
      fn vkGetInstanceProcAddr();
    }

   // load everything else with GetInstanceProcAddr,
   // since loader is not guaranteed to expose any other functions
  }
}
```

In this scenario, Rust will _not_ link against `libvulkan.so` unless the user calls `static()` at some point.

**Comment by Ralith** [comment]
> Especially on mobile devices, where driver updates are quite rare, many games prefer to try to dynamically load Vulkan, and fall back to OpenGL if that's not supported.

It is unfortunate for the usability of the library on desktop platforms to be compromised by mobile.

> In the end, loading the drivers on any platform will require dynamic linking.

Dynamic linking and dlopen are different things. The loader should be dynamically linked, and whatever dlopening the loader does itself is not a problem in the way that dlopening the loader is, because that dispatching is the entire function of the loader, and as such it will be configured as needed to do so painlessly.

> In this scenario, Rust will not link against libvulkan.so unless the user calls static() at some point.

Wow, since when is that a thing? That's pretty cool. I'm still not terribly happy about having to carry an `Entry` around, but it certainly solves the larger problems. However, note that "static" would be a misnomer there; at least in a linux environment, the loader should normally be dynamically linked, which is different than the dlopening that is currently mandated.

**Comment by Ralith** [comment]
> dlopen is pretty much the standard way of linking Vulkan and I don't see any negative effects.

To clarify, the negative effects are that the API has added friction due to the need for the `Entry` type, and the resulting program places nonstandard requirements on its runtime environment.

**Comment by MaikKlein** [comment]
I personally will only using runtime linking but I am not against optional dynamic or even static linking. It is just not a priority for me. PRs are welcome.
